docs/design_notes/module_lifecycle.md

```diff
@@ -0,0 +1,289 @@
+
+### Construction/initialization
+
+Now we want to construct and use the `MLP` Module:
+
+
+```python
+mlp = MLP(5, 3)
+variables = mlp.init(random.PRNGKey(0), x)
+y = mlp.apply(variables, x)
+```
+
+
+First, we construct an instance of `MLP` and pass the construction attributes. Note that construction here is different from what you might expect if you are not used to Functional Programming patterns. The `MLP` constructor does not actually create variables or any internal state whatsoever. It’s best to think of it as a specification or template of the Module that contains functionality but no data.
+
+Let’s first take a closer look at initialization. Surprisingly, there actually is no special initialization path in Flax. Calling `init` is just a special case of `apply` and we could also code it as:
+
+
+```python
+# equivalent to: variables = mlp.init(random.PRNGKey(0), x)
+_, variables = mlp.apply({}, x, rngs={"params": random.PRNGKey(0)}, mutable=True)
+```
+
+
+Thus, `init` is nothing more than a special case of `apply` where:
+
+1. We call a module without any initial variables (an empty dict)
+2. A PRNG generator named “params” is always passed for randomly initializing parameters.
+3. All variable collections are set to mutable (`mutable=True`). When a collection is mutable, existing variables can be updated and new variables can be created.
+
+### Lifecycle
+
+Thus, what remains is to actually understand what `.apply(...)` does. All the interesting stuff in a Module happens inside the `apply` method. We will refer to the various stages before and during `apply` as the “Module lifecycle”. We can summarize this lifecycle in a number of steps:
+
+
+1. We construct `mlp = MLP(5, 3)`, such that `mlp.hidden_size=5` and `mlp_out_size=3`
+2. Then, `mlp.apply`:
+    1. Makes a clone of `mlp`, let’s call it `mlp_copy`
+    2. Calls `mlp_copy.setup()`.
+    3. Returns the output of `mlp_copy.__call__()` and optionally the variable collections that were specified as mutable using `mutable=` kwarg.
+
+
+You might be surprised that the lifecycle includes cloning the Module instance. This will be explained later in the **Top-level Modules** section.
+
+### Variables
+
+The word “variable” is ubiquitous in programming and math. However, it’s important to have a good understanding of what that means in the context of JAX and Flax. Inside Flax modules, variables act like you expect from Python. They are initialized once, read, and perhaps even updated every so often. However, JAX has no concept of variables. Instead, values are stored in arrays similar to NumPy arrays  -- with one important difference: they are immutable.
+
+`init` and `apply` return the variables as a dictionary of ordinary JAX arrays. They are immutable and therefore really just a snapshot of the variables. Next time we call apply we can feed back the snapshot so the variables are again in the same state as where we finished on the last call to apply.
```

Comment:
I don't really get the "Next time" sentence. what does feeding back the snapshot mean?

Reply:
clarified